django/db/migrations/autodetector.py

```diff
@@ -833,11 +833,26 @@ def generate_removed_fields(self):
             self._generate_removed_field(app_label, model_name, field_name)
 
     def _generate_removed_field(self, app_label, model_name, field_name):
+        field = self.old_apps.get_model(app_label, model_name)._meta.get_field(field_name)
+
+        # You can't just remove NOT NULL fields with no default or fields
+        # which don't allow empty strings as default because the reverse
+        # migration will need a default value to populate existing records in
+        # the database while adding the field back.
+        temp_default = models.NOT_PROVIDED
+        time_fields = (models.DateField, models.DateTimeField, models.TimeField)
+        if (not field.null and not field.has_default() and
+                not field.many_to_many and
+                not (field.blank and field.empty_strings_allowed) and
+                not (isinstance(field, time_fields) and field.auto_now)):
```

Comment:
It looks like you also need [`or field.auto_now_add` here](https://github.com/django/django/blob/de4265e082bcce45ef6fb0929a44205463ba552e/django/db/backends/base/schema.py#L205). There's a lot of duplication between this conditional and the ones in `BaseDatabaseSchemaEditor.effective_default` do you think it could be worth refactoring both by adding an additional `effective_default` property to `Field`?